devito/types/dense.py

```diff
@@ -1311,3 +1313,199 @@ def parent(self):
+                names += d.name
+            return names
+
+        nonseparated = tuple(d for d in grid.dimensions if d not in separated)
+        if len(nonseparated) > 0:
+            func_list += [Function(name=name+'_'+names(nonseparated),
+                                   dimensions=nonseparated, grid=grid,
+                                   shape=tuple(grid.dimension_map[d].loc for d in nonseparated),
+                                   space_order=so, **kwargs)]
+
+        expr = sep_op(*func_list)
+        new_obj = Differentiable.__new__(cls, expr)
+
+        dims = tuple([d for d in separated] + [nonseparated])
+        new_obj._subfunctions = {d: f for d, f in zip(dims, func_list)}
+        new_obj._dims = dims
+        new_obj._expr = expr
+
+        return new_obj
+
+    def __repr__(self):
+        return self.expr.__repr__()
+
+    __str__ = __repr__
+
+    def from_expr(self, expr):
+        new_obj = Differentiable.__new__(SeparableFunction, expr)
+        new_obj._subfunctions = {d: f for d, f in zip(self.dims, expr.args)}
+        new_obj._dims = self.dims
+        new_obj._expr = expr
+
+        return new_obj
+
+    def func(self, *args, **kwargs):
+        return self.from_expr(self.expr.func(*args, **kwargs))
+
+    @property
+    def expr(self):
+        return self._expr
+
+    @property
+    def dims(self):
+        return self._dims
+
+    @property
+    def subfunctions(self):
+        return self._subfunctions
+
+    def __getitem__(self, dim):
+        return self.subfunctions[dim]
+
+    @property
+    def data(self):
+        return tuple(f.data for f in self.subfunctions.value)
+
+    @property
+    def evaluate(self):
+        return self
+
+    def subs(self, rules):
+        return self.from_expr(self.expr.subs(rules))
+
+    def xreplace(self, rules):
+        return self.from_expr(self.expr.xreplace(rules))
+
+
+class SeparableTimeFunction(SeparableFunction):
+    """
+    A TimeFunction separable in its Dimensions, ie f(t,x,y,z) = f(t)*f(x)*f(y)*f(z)
+
+    Take the same parameters as a TimeFunction plus
+
+    Parameters
+    ----------
+    sep: str
+        Type of separation as a product or sum of Function, 'sum' or 'prod'.
+    separated: Dimension or tuple of Dimensions
+        Dimension in which the TimeFunction is separable.
+
+    Examples
+    --------
+    Creation
+
+    >>> from devito import Grid, TimeFunction
+    >>> grid = Grid(shape=(4, 4, 5))
+    >>> x, y, z = grid.dimensions
+    >>> t = grid.stepping_dim
+    >>> SeparableTimeFunction(name='a', grid=grid, space_order=8, time_order=2)
+    a_t(t)*a_x(x)*a_y(y)*a_z(z)
+    >>> SeparableTimeFunction(name='b', grid=grid, sep='sum')
+    b_t(t) + b_x(x) + b_y(y) + b_z(z)
+    >>> SeparableTimeFunction(name='c', grid=grid, sep='sum', separated=t)
+    c_t(t) + c_xyz(x, y, z)
+    >>> SeparableTimeFunction(name='d', grid=grid, separated=(t, x))
+    d_t(t)*d_x(x)*d_yz(y, z)
+    """
+    def __new__(cls, *args, **kwargs):
```

Comment:
I also believe that much of the code in this ``__new__`` can be performed by the superclass via ``super``. We should really aim to achieve this, as currently there's too much code duplication